File: Python/Binary_Exponentiation.py
def power(x, y, p):
    res = 1
 
    while (y > 0):
        if ((y & 1) != 0):
            res = res * x
        y = y >> 1
        x = x * x
 
    return res % p
 
x = 2
y = 5
p = 13
print("Power is ", power(x, y, p))

File: Cpp/BinaryExponentiation.cpp
#include <bits/stdc++.h>
using namespace std;

long long power(long long num, long long pw)
{
    long long res = 1;
    while (pw)
    {
        if (pw & 1)
            res *= num;
        num *= num;
        pw >>= 1;
    }
    return res;
}

const int MOD = 1e9 + 7;
long long mod_pow(long long num, long long pw)
{
    long long res = 1;
    while (pw)
    {
        if (pw & 1)
        {
            res *= num;
            res %= MOD;
        }
        num *= num;
        num %= MOD;
        pw >>= 1;
    }
    return res;
}

int main()
{
    long long num, pw;
    cin >> num >> pw;
    cout << power(num, pw) << endl;
    cout << mod_pow(num, pw) << endl;
    return 0;
}